### DISCUSSION TITLE
step 3 timing out/connection closed

### DISCUSSION THREAD

**Discussion by salvesen** [timdorr/tesla-api#320] [Q&A]
Not sure what is going on today. This has been working for weeks for me now and this morning I am just getting timeouts on step 1. I tried adding the optional hint, no change. Could this be due to the WAF? I do not have a fix for this in my code, but since https://auth.tesla.com/oauth2/v3/authorize is just timing out when trying to download data I have no idea how to bypass it. Any ideas? 

`Exception getLoginPageHTML: System.Net.WebException: The operation has timed out`

**Comment by salvesen** [comment]
While testing I can see that I sometime manage to download step 1 but then it fails on step 2. Not sure what is going on here..

**Comment by corsair** [comment]
I am also now experiencing this issue on [Teslascope](https://teslascope.com) and no accounts are able to authenticate. It appears to be failing regardless of the codebase (tried PHP and Python).

This seems troubling. Is this impacting the main repository? @timdorr

**Comment by nilathedragon** [comment]
It seems if you send no headers at all, it suddenly works. I now only send the User-Agent in my script and no additional headers.: 

`    headers = {
        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Safari/605.1.15"
    }`

**Reply by salvesen** [reply]
Not sure what you mean here, step 1 does not require any headers?

**Reply by nilathedragon** [reply]
I was going from this script https://github.com/enode-engineering/tesla-oauth2 where there are headers being sent in every step... Thats what did the trick for me

**Comment by salvesen** [comment]
Still struggeling here, today step 1 & 2 looks to be ok. But on step 3 when getting the code the redirect message has changed. There is no more "code" in it. 

`redirect_uri=https://auth.tesla.com/void/callback&response_type=code&scope=openid%20email%20offline_access&state=123`

what the fudge is going on..

**Reply by salvesen** [reply]
My bad, this was due to a missing header on my side when troubleshooting. But today, step 3 is timing out. Everything else before this is ok and is done as per https://tesla-api.timdorr.com/api-basics/authentication#refreshing-an-access-token
I also tested my "old" code before i started to work on this and the same issue. timeouts move to step 3. 

Anyone have  any idea on what is going on? @timdorr ?

**Reply by salvesen** [reply]
`Exception exchangeAuthCode: System.Net.WebException: The underlying connection was closed: A connection that was expected to be kept alive was closed by the server. ---> 
System.IO.IOException: Unable to read data from the transport connection: A connection attempt failed because the connected party did not properly respond after a period of time, 
or established connection failed because connected host has failed to respond. ---> System.Net.Sockets.SocketException: A connection attempt failed because the connected party did not properly respond after a period of time,
 or established connection failed because connected host has failed to respond`

**Reply by tomhollander** [reply]
Yes this is what I'm seeing too. Very occasionally it works, which only serves to make me wonder what I did right, but then it goes right back to timing out again.

**Reply by chrm-software** [reply]
Same here. Step 3 just timed out. No error message, no answer at all.
I'm wondering, what kind of API are services like TeslaFi using? They seems to have a working solution.

**Reply by salvesen** [reply]
yeah this is def very wierd. Maybe it is due to all the traffic tesla is getting now as 3rd party apps have resolved their auth? I dont know, but it is really anyoing.

**Comment by fkhera** [comment]
I am timing out getting the access token now so appears Tesla changed something yesterday times out at at auth page login.
I have not removed headers yet hoping to avoid code changes to deploy :
https://github.com/fkhera/powerwallCloud/blob/master/powerwallBackup.py#L252

**Comment by fkhera** [comment]
I ended up just stripping headers, cause its 11pm Sunday night and tomorrow I have to ensure my powerwalls as well as others are programmed correctly.  Really hoping this is the final fix.  If they block no header requests, let us know, cause then I will have to re-update the code as well.

Latest commit to patch:
https://github.com/fkhera/powerwallCloud/commit/393e1fc5ebd0547a26848aba680469e227a36855

**Reply by salvesen** [reply]
Did this really work? I see you only changed the user-agent headers. And they are not even needed in the example given: https://tesla-api.timdorr.com/api-basics/authentication#refreshing-an-access-token

I am not getting this to work without user-agent headers. Can you specify what you are skipping in the example, please?

**Reply by chrm-software** [reply]
Clearing der UA-header did not work for me. But what I found out is that the whole process works fine if you wakeup the car first using teslas stock app.
This is really stupid, because I use the API to even wake up the car. I think there is something broken in Teslas system...

**Reply by salvesen** [reply]
@chrm-software yeah something is really off. Waking up does not solve it for me, still stuck on step 3 :/

**Reply by salvesen** [reply]
It worked one time, but still very unstable!

**Reply by tomhollander** [reply]
I was able to get past the timeout in step 3 by adding the following headers to the call:
```
Accept: application/json
Accept-Encoding: gzip, deflate
Content-Type: application/json; charset=utf-8
Connection: keep-alive
```

**Reply by chrm-software** [reply]
Did not work for me. Setting the "Accept"-header generates: server replied: Bad Request

**Reply by chrm-software** [reply]
Do you use a real UserAgent? Which one?

**Reply by salvesen** [reply]
Also not working for me, I still have the same issue on my side: System.Net.WebException: The underlying connection was closed:

**Reply by chrm-software** [reply]
Setting the Accept-Header to json/text/whatever generates this answer:
{"error":"invalid_request","error_description":"Internal server error","error_uri":"https://auth.tesla.com/error/reference <RANDOM-NUMBERS-HERE>"}

**Reply by chrm-software** [reply]
So for me it only works if I wakeup the car from sleeping - but only once.

**Reply by salvesen** [reply]
@chrm-software  I also got it working one time today, but all tries after fail's with the timeouts. So its def not something that will work in the long run.

**Reply by tomhollander** [reply]
@salvesen Can you try my library at https://github.com/tomhollander/TeslaAuth, since you're using .NET?
The changes I did to fix it after the recent changes are here: https://github.com/tomhollander/TeslaAuth/commit/b9f008d53e497cc3d7783bb093a3f144644ba17c

**Reply by salvesen** [reply]
@tomhollander yes this is working! Thanks, ill try to sort my code out based on your changes

**Reply by BJReplay** [reply]
@tomhollander Thanks for this, got my login working again.

**Reply by chrm-software** [reply]
@tomhollander Thanks! Setting "Accept" header to:  "application/json" did the trick for me also - earlier today it was just my mistake. There is no more error message form the service. I'm able to get the state from sleeping car and make a wakeup call.

**Reply by salvesen** [reply]
@tomhollander I got mine working again by using your ExchangeCodeForBearerTokenAsync function modified to fit within my code. Thanks a lot and hope you dont mind me using some of the code in that function. 

I have no idea what is wrong with my original code in sep 3, but this did the trick for me. 

Thanks a lot!

**Reply by dewoodruff** [reply]
Just sharing my experience. I also started getting timeouts at step 3 with my nodejs app in the last few days. Adding the 
`"Connection": "keep-alive"`
header, in addition to a mobile user agent I was already sending, worked for a bit but is timing out again. Adding Accept, Accept-Encoding, and Content-Type caused a 400 bad request error.

**Reply by chrm-software** [reply]
Same for me again... Worked fine for four days but now: 400 Bad Request at this step or timeout with empty headers.

**Reply by salvesen** [reply]
> Same for me again... Worked fine for four days but now: 400 Bad Request at this step or timeout with empty headers.

yeah timing out again here to. Now on step1 again. Also tried the @tomhollander example. and it is also failing :/

**Reply by tomhollander** [reply]
I just updated my library - I think it's working again.
It seems that most/all of the calls now expect:
```
Accept: application/json
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
```

**Reply by salvesen** [reply]
> I just updated my library - I think it's working again.
> It seems that most/all of the calls now expect:
> 
> ```
> Accept: application/json
> Accept-Encoding: gzip, deflate, br
> Connection: keep-alive
> ```

Just tested it and it is working! unfortunately for me my lib is not. Tried adding all of that to my "step1" and the html page that is returned is just jibberish. Parts of it: 
`▼       í]ÝoãHr⌂÷_ÑÇ♣▬▲`s↕%Ù'g,%¶Ç³3;▼;·zÙÝìå ´É-H<▼Z6i?½↓ yI^♫ÈSroI?C'╬ yHz6⌂Jvƒä?ÿ!UÝMS▼²lJzÃb³?G">ÉÓp¿¥.mma·YPJ>å_¥^Ì↔↕ð"'"M♣¡¶xÉSÛe±àÉÐH"        ↔↑¤]p"dN'ò|h∟EaÂÃ"¾┼¡↓ÄVgC#áo"▬♫ñaÎf§-/è>♥z¶♣s-x§~'ÑÓã!w¦ü¾íÆQÀ╬V?AÈ Á8÷øÅ<S"☻Í.ç$îÐáçzÍ©<¹O¼ÐK<æSa3▼ØÜ'☺{ë♣i°hH♣?å↓ƒ◄♀Û÷IÆTN¼dhGç<-Â·ö}/o'~ûCC$->↨.ç Þ?ùdh´D☻:Ø-[^Ö4J\▬ÐÓ(sTp®┼Þ?8àZ-♠ÍÉ♥P8`¾ys┼ZÏ>Ó♂×♂"       \rÒ­}aÇÞ<!"¶↨}ÏDË³♠!Lw&
iÎ"1Úo©þ£"®Ê4u⌂4wmJgìo©Vƒ"QhCûî¤ÓéwºƒSnq>YÖiwÏ6Ð☺WI♥0♣"9Îñ9∟<÷♦~♠?·?GY¼Ð&û<b♫wOûd'┼vâEá6Ç®÷È×'!l>÷/▼±xö"r8ùðÃrÃö½╬[ºã9<Á?"`S.È?|ý.¿2%b²?-çq4←Ì»-¤?∟↑d∟Gñ♂um;ëó3$úù}I♂Â`


How can we keep up with this, its so unstable.

**Reply by tomhollander** [reply]
@salvesen That'll be the gzipped response. It's not enough to add the headers; you also need to decode the payload. Hopefully your chosen platform can do this for you if you choose the right options.

**Reply by chrm-software** [reply]
For me it works with "Accept-Encoding: deflate, br" too. So no need to accept gzipped response.

**Reply by salvesen** [reply]
> @salvesen That'll be the gzipped response. It's not enough to add the headers; you also need to decode the payload. Hopefully your chosen platform can do this for you if you choose the right options.

ah that makes sense! Got it decompressed now and looks good. But still stuck on step 3. Ill have a look there again :)

**Reply by salvesen** [reply]
back up and running thanks to @tomhollander tips! Thanks again!

**Comment by fkhera** [comment]
So far the stripped headers is still working, I guess we keep with it for a while

**Reply by dewoodruff** [reply]
Thanks for sharing your experience. I wrote an app in nodejs and have also been getting timeouts the last few days. I'm finding that I'm able to auth once, then I get a timeout. If I change ANY header value, even just the UA to a different UA, I'm able to immediately authenticate again. Maybe they're rate limiting auth requests base on fixed header values.

**Comment by fkhera** [comment]
MAke sure to remove usage agent from headers

**Comment by fkhera** [comment]
So far do we thinking removing the headers is best way to fix this issue ? I have to fix 2 customers devices cause they do not give remote access rasberry pi

**Reply by tomhollander** [reply]
Not so much removing headers, but having (only) the expected headers.

**Comment by pgalbavy** [comment]
Long time listener, first time caller... :-)

I have been hitting a timeout on the .../token call since doing this. I have had problems sometimes when the JSON body is wrong, but other times even sending no body and it just times out. Looks like the WAF is expecting very specific things. I have tried all the above header adds and removes suggested, no consistent behaviour. I have also tried two different source IPs but only have my own account to test with (which works fine on web and real app).

Is there any reliance on earlier stages (session cookie etc.) working or is it purely the code and code_verifier?

Last replies here over a week ago, so just wondering if it's still working for others properly? 

Update: On a hunch I tried forcing TLS1.3 and got a different response. Going to now try reworking my real test
Update 2: Actually, it seems to be something else, but at least now I'm getting an "invalid code_verifier" error
Update 3: Datapoint, explicitly setting the SNI to auth.tesla.com seems to do something, not relying on the underlying API (I'm using node-red)

**Reply by salvesen** [reply]
maybe this lib can help you out, https://github.com/tomhollander/TeslaAuth just tested and it works. My app also works, but i dont have it on git.

**Reply by pgalbavy** [reply]
Thanks, I read through that (and many others) very carefully and I cannot see what I am doing differently.

I have reduced the problem down further and it "feels" very much like the HTTP connection and headers and not the JSON, but if I supply intentionally incorrect code or code_verifier I can trigger an error rather than a timeout.

I will find a project that builds a command than I can try to run, rather than PHP/Node.js for web use, and see what I see.

**Reply by pgalbavy** [reply]
OK, I rewrote my impression of the code based on @timdorr docs in golang and it worked without issues. Must be somethinig Node Red is doing inside it's http request nodes.

**Reply by pgalbavy** [reply]
... and then today, while trying to clear up my hacky test code it seems they now need user-agent headers. Works yesterday, 403 Forbidden today. Sigh.

**Reply by salvesen** [reply]
@pgalbavy just tested mine and previously mentioned lib. Working as it should, no user-agent headers.

**Reply by dewoodruff** [reply]
My hypothesis is requests with the same exact headers from the same source IP are rate limited. Oauth isn't supposed to be used to reauthenticate every single request for data. You're supposed to auth, grab the access keys once, and then reuse them for a long time. I suspect they're rate limiting to prevent people following bad authentication practices.

When you get an error, try changing any header value, like user agent. Does it immediately start working again?

**Reply by pgalbavy** [reply]
You are probably right. I have not yet started cacheing the bearer token etc. as I am still cleaning the code - but I guess adding some randomisation to the user-agent header may help if your theory proves right.

**Comment by PatrickSchmidtSE** [comment]
"Last replies here over a week ago, so just wondering if it's still working for others properly?"
It worked over 1 week with me, but now since 2 hours everything is back to not working again :/ frustrating!!

**Reply by salvesen** [reply]
Just tested mine now, works fine for now.

**Reply by salvesen** [reply]
Also tested @tomhollander liberary now, works also. Maybe this can help you out? https://github.com/tomhollander/TeslaAuth

**Reply by PatrickSchmidtSE** [reply]
Yeah was a one time error, sorry to make trouble :)

**Reply by salvesen** [reply]
@SearchForTheCode glad to hear it:)
